File: experiments/identity_paths/grouping.py
"""Leiden consensus followed by deterministic complete-link consolidation."""

from __future__ import annotations

from itertools import combinations

import igraph
import leidenalg
import numpy as np

from experiments.core_discovery.candidates import ownership
from experiments.identity_paths.config import GROUP_JACCARD, SEEDS


def similarities(
    candidates: list[dict], matrix: np.ndarray, items: tuple
) -> np.ndarray:
    index = {item: column for column, item in enumerate(items)}
    masks = [
        ownership(matrix, tuple(index[item] for item in candidate["items"]))
        for candidate in candidates
    ]
    result = np.eye(len(candidates))
    for first, second in combinations(range(len(candidates)), 2):
        common = set(candidates[first]["items"]) & set(candidates[second]["items"])
        if len(common) < 2:
            continue
        jaccard = (masks[first] & masks[second]).sum() / max(
            1, int((masks[first] | masks[second]).sum())
        )
        if jaccard >= GROUP_JACCARD:
            result[first, second] = result[second, first] = jaccard
    return result


def merge_complete(weights: np.ndarray, eligible: np.ndarray) -> list[list[int]]:
    groups = [(index,) for index in range(len(weights))]
    while True:
        choices = []
        for first, second in combinations(range(len(groups)), 2):
            block = np.ix_(groups[first], groups[second])
            if eligible[block].all():
                merged = tuple(sorted((*groups[first], *groups[second])))
                choices.append((-float(weights[block].mean()), merged, first, second))
        if not choices:
            return [list(group) for group in sorted(groups)]
        _, merged, first, second = min(choices)
        groups = [
            group for index, group in enumerate(groups) if index not in {first, second}
        ]
        groups.append(merged)
        groups.sort()


def consolidate(candidates: list[dict], matrix: np.ndarray, items: tuple) -> dict:
    weights = similarities(candidates, matrix, items)
    first, second = np.nonzero(np.triu(weights > 0, 1))
    graph = igraph.Graph(
        n=len(candidates), edges=list(zip(first.tolist(), second.tolist(), strict=True))
    )
    assignments = []
    for seed in SEEDS:
        if not len(first):
            membership = list(range(len(candidates)))
        else:
            partition = leidenalg.find_partition(
                graph,
                leidenalg.RBConfigurationVertexPartition,
                weights=weights[first, second].tolist(),
                resolution_parameter=1,
                n_iterations=10,
                seed=seed,
            )
            membership = list(partition.membership)
        assignments.append({"seed": seed, "membership": membership})
    votes = np.zeros_like(weights, dtype=int)
    for assignment in assignments:
        labels = np.asarray(assignment["membership"])
        votes += labels[:, None] == labels[None, :]
    groups = merge_complete(weights, (weights > 0) & (votes >= 2))
    return {
        "groups": groups,
        "seeds": assignments,
        "edges": [
            {
                "first": int(left),
                "second": int(right),
                "jaccard": float(weights[left, right]),
                "votes": int(votes[left, right]),
            }
            for left, right in zip(first, second, strict=True)
        ],
        "candidate_order": [row["items"] for row in candidates],
    }

File: crates/deadlock-analysis/src/grouping.rs
use deadlock_data::{Error, Result, count_as_f64, count_ratio};
use leiden_rs::{GraphDataBuilder, Leiden, LeidenConfig, QualityType};
use serde_json::{Value, json};

use crate::discovery_data::DiscoveryData;
use crate::mining::Candidate;

pub fn group_candidates(candidates: &[Candidate], data: &DiscoveryData) -> Result<Value> {
    let weights = ownership_similarities(candidates, data)?;
    group_weights(candidates, &weights, true)
}

pub fn group_item_candidates(candidates: &[Candidate]) -> Result<Value> {
    let mut weights = vec![vec![0.0; candidates.len()]; candidates.len()];
    for first in 0..candidates.len() {
        weights[first][first] = 1.0;
        for second in first + 1..candidates.len() {
            let shared = candidates[first]
                .items
                .iter()
                .filter(|item| candidates[second].items.contains(item))
                .count();
            let union = candidates[first].items.len() + candidates[second].items.len() - shared;
            let score = count_ratio(u64::try_from(shared)?, u64::try_from(union.max(1))?)?;
            if shared >= 2 && score >= 0.5 {
                weights[first][second] = score;
                weights[second][first] = score;
            }
        }
    }
    group_weights(candidates, &weights, false)
}

fn group_weights(
    candidates: &[Candidate],
    weights: &[Vec<f64>],
    positive_edges: bool,
) -> Result<Value> {
    let size = candidates.len();
    let mut builder = GraphDataBuilder::new(size);
    let mut edges = Vec::new();
    for (first, row) in weights.iter().enumerate() {
        for (second, weight) in row.iter().enumerate().skip(first + 1) {
            if *weight > 0.0 {
                builder
                    .add_edge(first, second, *weight)
                    .map_err(|error| grouping_error(&error))?;
                edges.push((first, second));
            }
        }
    }
    let graph = builder.build().map_err(|error| grouping_error(&error))?;
    let mut assignments = Vec::new();
    let mut votes = vec![vec![0_u8; size]; size];
    for seed in [42, 43, 44] {
        let membership = if edges.is_empty() {
            (0..size).collect::<Vec<_>>()
        } else {
            Leiden::new(LeidenConfig {
                max_iterations: 10,
                min_iterations: 10,
                quality: QualityType::RBConfiguration,
                resolution: 1.0,
                seed: Some(seed),
                ..LeidenConfig::default()
            })
            .run(&graph)
            .map_err(|error| grouping_error(&error))?
            .partition
            .as_slice()
            .to_vec()
        };
        for first in 0..size {
            for second in 0..size {
                votes[first][second] += u8::from(membership[first] == membership[second]);
            }
        }
        assignments.push(json!({"seed":seed,"membership":membership}));
    }
    let groups = merge_groups(weights, &votes, positive_edges)?;
    Ok(
        json!({"groups":groups,"seeds":assignments,"edges":edges.iter().map(|(first,second)|json!({"first":first,"second":second,"jaccard":weights[*first][*second],"votes":votes[*first][*second]})).collect::<Vec<_>>(),
        "candidate_order":candidates.iter().map(|candidate|&candidate.items).collect::<Vec<_>>()}),
    )
}

fn ownership_similarities(candidates: &[Candidate], data: &DiscoveryData) -> Result<Vec<Vec<f64>>> {
    let masks = candidates
        .iter()
        .map(|candidate| {
            let columns = data.columns(&candidate.items)?;
            Ok(data
                .fold_rows("discovery")
                .map(|row| data.owns(row, &columns))
                .collect::<Vec<_>>())
        })
        .collect::<Result<Vec<_>>>()?;
    let mut weights = vec![vec![0.0; candidates.len()]; candidates.len()];
    for first in 0..candidates.len() {
        weights[first][first] = 1.0;
        for second in first + 1..candidates.len() {
            if candidates[first]
                .items
                .iter()
                .filter(|item| candidates[second].items.contains(item))
                .count()
                < 2
            {
                continue;
            }
            let intersection = masks[first]
                .iter()
                .zip(&masks[second])
                .filter(|(left, right)| **left && **right)
                .count();
            let union = masks[first]
                .iter()
                .zip(&masks[second])
                .filter(|(left, right)| **left || **right)
                .count();
            let weight = count_ratio(u64::try_from(intersection)?, u64::try_from(union.max(1))?)?;
            if weight >= 0.7 {
                weights[first][second] = weight;
                weights[second][first] = weight;
            }
        }
    }
    Ok(weights)
}

fn merge_groups(
    weights: &[Vec<f64>],
    votes: &[Vec<u8>],
    positive_edges: bool,
) -> Result<Vec<Vec<usize>>> {
    let mut groups = (0..weights.len())
        .map(|index| vec![index])
        .collect::<Vec<_>>();
    loop {
        let mut choices = Vec::new();
        for first in 0..groups.len() {
            for second in first + 1..groups.len() {
                let pairs = groups[first]
                    .iter()
                    .flat_map(|left| groups[second].iter().map(move |right| (*left, *right)))
                    .collect::<Vec<_>>();
                if !pairs.iter().all(|(left, right)| {
                    votes[*left][*right] >= 2 && (!positive_edges || weights[*left][*right] > 0.0)
                }) {
                    continue;
                }
                let mean = pairs
                    .iter()
                    .map(|(left, right)| weights[*left][*right])
                    .sum::<f64>()
                    / count_as_f64(u64::try_from(pairs.len())?)?;
                let mut merged = groups[first]
                    .iter()
                    .chain(&groups[second])
                    .copied()
                    .collect::<Vec<_>>();
                merged.sort_unstable();
                choices.push((mean, merged, first, second));
            }
        }
        choices.sort_by(|left, right| {
            right
                .0
                .total_cmp(&left.0)
                .then_with(|| left.1.cmp(&right.1))
                .then(left.2.cmp(&right.2))
                .then(left.3.cmp(&right.3))
        });
        let Some((_, merged, first, second)) = choices.into_iter().next() else {
            return Ok(groups);
        };
        groups.remove(second);
        groups.remove(first);
        groups.push(merged);
        groups.sort();
    }
}

fn grouping_error(error: &leiden_rs::error::LeidenError) -> Error {
    Error::new(format!("Core grouping failed: {error}"))
}
